### ISSUE TITLE
Question: Implementing Retry Budget

### ISSUE THREAD

**Issue by bbkgh** [microbus-io/fabric#47]
Hi,

I’ve been reading your documentation to gain a better understanding of your software architecture, which I find quite interesting. Additionally, I came across this blog post today: [Good Retry, Bad Retry: An Incident Story](https://medium.com/yandex/good-retry-bad-retry-an-incident-story-648072d3cee6).

I have a few questions regarding the Retry mechanism in your architecture:

How is the Retry mechanism implemented?
What happens to messages if there is no subscriber service?
Is there a TTL (Time-to-Live) for messages, or can the caller retry sending messages with sth like Retry Budget?

**Comment by bw19** [comment]
By and large, Microbus does **not** include a retry mechanism.

For one, retries can indeed cause retry storms that make recovery very difficult. That is mentioned in the blog post you linked to.

Second, retries are an arguably reasonable strategy when upstream services "push" requests to a downstream services. By "push" I mean that the request is sent to a particular downstream service by its hostname, as is the case with HTTP. Because the upstream can never be sure that the downstream is there, it can end up getting a 404. A retry allows it to recover by targeting another downstream by a different hostname. If host x is not there, maybe host y is.

Microbus operates on a "pull" model. The upstream publishes the message to a queue (on the messaging bus), and a downstream service consumes it. There is no chance of a 404, unless there are no downstream consumers at all, in which case a retry will not help.

There is one edge case worth noting. If the downstream disconnects from the bus, but the bus is unaware of this disconnect, messages might end up getting queued on the bus with no one to consume them, and ultimately get dropped. In this case, the upstream will receive a 404 timeout error and can decide how to react. Because this is an edge case, I do not recommend a retry. It is better to error out until the bus eventually recognizes the dropped connection via its ping mechanism. Retries bring risks, which is why they are not implemented out of the box.

Bottom line: Retries are not a recommended strategy in Microbus. If you implement them at the app level, use them with extreme caution.
